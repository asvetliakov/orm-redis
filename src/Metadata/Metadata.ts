/**
 * Metadata key to mark class to be stored as redis hash
 */
export const REDIS_ENTITY = Symbol("Redis Hash");

/**
 * Metadata key to store redis properties metadatas
 */
export const REDIS_PROPERTIES = Symbol("Redis properties");

/**
 * Metadata key to keep serialized initial value of property from redis. For collections this will be link to the corresponding hashmap/set
 */
export const REDIS_VALUE = Symbol("Redis initial value");

/**
 * Metadata key to keep the initial state of redis collections (map/set)
 */
export const REDIS_COLLECTION_VALUE = Symbol("Redis initial collection value");


export type PropertyMetadata = SimplePropertyMetadata | RelationPropertyMetadata;

/**
 * Hash simple property metadata
 * 
 * @export
 * @interface SimplePropertyMetadata
 */
export interface SimplePropertyMetadata {
    /**
     * Property name in class
     */
    propertyName: string;
    /**
     * Property name stored in redis
     */
    propertyRedisName: string;
    /**
     * Property type
     */
    propertyType: any;
    /**
     * True if property must be identify column
     */
    isIdentifyColumn: boolean;
    /**
     * Relation flag
     */
    isRelation: false;
}

/**
 * Relation options
 * 
 * @export
 * @interface RelationOptions
 */
export interface RelationOptions {
    /**
     * Property name in redis
     */
    propertyName?: string;
    /**
     * True to automatically create new hash for relation
     */
    cascadeInsert?: boolean;
    /**
     * True to automatically update corresponding relation hash
     */
    cascadeUpdate?: boolean;
}

/**
 * Hash relation property metadata
 * 
 * @export
 * @interface RelationPropertyMetadata
 */
export interface RelationPropertyMetadata {
    /**
     * Property name in class
     */
    propertyName: string;
    /**
     * Property name stored in redis
     */
    propertyRedisName: string;
    /**
     * Property type.
     * For set property relation it will be Set itself
     * For single property relation it will be equal to relationType
     */
    propertyType: any;
    /**
     * Relation flag
     */
    isRelation: true;
    /**
     * Relation type. Must be type of related entity
     */
    relationType: any;
    /**
     * Relation options
     */
    relationOptions: RelationOptions;
}

/**
 * Return true if given object is redis entity like
 * 
 * @export
 * @param entity 
 * @returns 
 */
export function isRedisEntity(entity: object): boolean {
    if (typeof entity !== "object" || entity === null) {
        return false;
    }
    const metadata = Reflect.getMetadata(REDIS_ENTITY, entity.constructor);
    return !!metadata;
}

/**
 * Return entity name
 * 
 * @export
 * @param entityOrEntityClass 
 * @returns 
 */
export function getEntityName(entityOrEntityClass: object | Function): string | undefined {
    const metadata = Reflect.getMetadata(REDIS_ENTITY, typeof entityOrEntityClass === "function" ? entityOrEntityClass : entityOrEntityClass.constructor);
    return metadata;
}

/**
 * Return entity id
 * 
 * @export
 * @param entity 
 * @returns 
 */
export function getEntityId(entity: { [key: string]: any }): string | number | undefined {
    if (!isRedisEntity(entity)) {
        return undefined;
    }
    const metadatas: PropertyMetadata[]  = Reflect.getMetadata(REDIS_PROPERTIES, entity.constructor);
    const idMetadata = metadatas && metadatas.find(val => !val.isRelation && val.isIdentifyColumn);
    if (!idMetadata) {
        return undefined;
    }
    if (typeof entity[idMetadata.propertyName] === "number" || typeof entity[idMetadata.propertyName] === "string") {
        return entity[idMetadata.propertyName];
    }
    return undefined;
}

/**
 * Returns redis properties metadata
 * 
 * @export
 * @param entityOrEntityClass 
 * @returns 
 */
export function getEntityProperties(entityOrEntityClass: object | Function): PropertyMetadata[] | undefined {
    return Reflect.getMetadata(REDIS_PROPERTIES, typeof entityOrEntityClass === "function" ? entityOrEntityClass : entityOrEntityClass.constructor);
}

/**
 * Return full hash id
 * 
 * @export
 * @param entityOrEntityClass 
 * @param [hashId]
 * @returns 
 */
export function getEntityFullId(entityOrEntityClass: object | Function, hashId?: string | number): string | undefined {
    const name = getEntityName(entityOrEntityClass);
    const id = typeof hashId !== "undefined"
        ? hashId
        : typeof entityOrEntityClass === "object"
            ? getEntityId(entityOrEntityClass)
            : undefined;        
    if (name && typeof id !== "undefined") {
        return `e:${name}:${id}`;
    }
}