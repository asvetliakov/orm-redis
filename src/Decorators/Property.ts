import { LazyMap } from "../collections/LazyMap";
import { LazySet } from "../collections/LazySet";
import { PropertyMetadata, REDIS_PROPERTIES } from "../Metadata/Metadata";

export type ValidType =
    typeof Number |
    typeof String |
    typeof Boolean |
    typeof Object |
    typeof Array |
    typeof Date |
    typeof Map |
    typeof LazyMap |
    typeof LazySet |
    typeof Set;


/**
 * Defines redis hash property
 * 
 * @export
 * @param [type] 
 * @returns 
 */
export function Property(type?: ValidType): PropertyDecorator;
/**
 * Defines redis hash property
 * 
 * @export
 * @param name Redis name for property
 * @param [type] 
 * @returns 
 */
export function Property(name: string, type?: ValidType): PropertyDecorator; 

/**
 * Defines redis hash property
 * 
 * @export
 * @param [name] Property name to store in redis
 * @returns 
 */
export function Property(nameOrType?: string | ValidType, type?: ValidType): PropertyDecorator {
    return function (target: Object, propertyKey: string): void {
        const designType: any = Reflect.getMetadata("design:type", target, propertyKey);
        const name = typeof nameOrType === "string" ? nameOrType : propertyKey;
        const propertyType = typeof nameOrType === "function"
            ? nameOrType
            : type
                ? type
                : designType;        

        const properties: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, target.constructor) || [];
        properties.push({
            propertyName: propertyKey,
            propertyRedisName: name,
            isIdentifyColumn: false,
            propertyType: propertyType,
            isRelation: false,
        });
        Reflect.defineMetadata(REDIS_PROPERTIES, properties, target.constructor);
    };
}