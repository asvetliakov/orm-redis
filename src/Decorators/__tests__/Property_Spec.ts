import { PropertyMetadata, REDIS_PROPERTIES } from "../../Metadata/Metadata";
import { Property } from "../Property";

// class Internal { }

class Test { 
    @Property()
    public one: string;
    @Property("twoName")
    public two: number;
    @Property(Date)
    public three: Date;
    @Property()
    public four: Object;
    @Property("fiveName")
    public five: { id: string };
    @Property()
    public six: boolean;
    // @Property(Int8Array, "sevenName")
    // public seven: Int8Array;
    @Property(Set)
    public eight: Set<string>;
    @Property("nineName", Map)
    public nine: Map<any, any>;
    // @Property()
    // public ten: Internal;
    @Property()
    public eleven: string[];
    
    public nonRedis1: string;
    public nonRedis2: number;
}

it("Defines metadata", () => {
    const metadata: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, Test);
    expect(metadata).toHaveLength(9);
    const [one, two, three, four, five, six, /* seven, */ eight, nine, /* ten, */eleven] = metadata;
    expect(metadata).toMatchSnapshot();
    expect(one.propertyType).toBe(String);
    expect(two.propertyType).toBe(Number);
    expect(three.propertyType).toBe(Date);
    expect(four.propertyType).toBe(Object);
    expect(five.propertyType).toBe(Object);
    expect(six.propertyType).toBe(Boolean);
    // expect(seven.propertyType).toBe(Int8Array);
    expect(eight.propertyType).toBe(Set);
    expect(nine.propertyType).toBe(Map);
    // expect(ten.propertyType).toBe(Internal);
    expect(eleven.propertyType).toBe(Array);
});