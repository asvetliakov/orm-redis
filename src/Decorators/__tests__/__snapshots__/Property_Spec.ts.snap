// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Defines metadata 1`] = `
Array [
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "one",
    "propertyRedisName": "one",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "two",
    "propertyRedisName": "twoName",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "three",
    "propertyRedisName": "three",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "four",
    "propertyRedisName": "four",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "five",
    "propertyRedisName": "fiveName",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "six",
    "propertyRedisName": "six",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "eight",
    "propertyRedisName": "eight",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "nine",
    "propertyRedisName": "nineName",
    "propertyType": [Function],
  },
  Object {
    "isIdentifyColumn": false,
    "isRelation": false,
    "propertyName": "eleven",
    "propertyRedisName": "eleven",
    "propertyType": [Function],
  },
]
`;
